src/dcqc/filesystems/synapsefs.py

```diff
@@ -0,0 +1,640 @@
+            synapse = Synapse(**self.synapse_args)
+            synapse.login(authToken=self.auth_token)
+            self._local.synapse = synapse
+        # Clear the Synapse cache to ensure up-to-date
+        self._local.synapse.cache.purge(after_date=0)
+        return self._local.synapse
+
+    def is_synapse_id(self, text: str):
+        return self.SYNID_REGEX.fullmatch(text) is not None
+
+    def _process_root(self, root: Optional[str]):
+        if root is None or root == "":
+            return None
+
+        root_path = PurePosixPath(root)
+        num_root_parts = len(root_path.parts)
+        error_message = f"Root ({root}) must be `None` or start with a Synapse ID."
+
+        if num_root_parts == 1:
+            if not self.is_synapse_id(root):
+                raise CreateFailed(error_message)
+        else:  # num_root_parts > 1
+            starting_entity, _, path = root.partition(self.DELIMITER)
+            if not self.is_synapse_id(starting_entity):
+                raise CreateFailed(error_message)
+            root = self._path_to_synapse_id(path, starting_entity)
+
+        # Ensure that the root is not a file
+        root_entity = self.synapse.get(root, downloadFile=False)
+        if not is_container(root_entity):
+            message = f"Root ({root}) must resolve to a project or folder."
+            raise CreateFailed(message)
+
+        return root
+
+    def _path_to_synapse_id(
+        self, path: str, starting_entity: Optional[str] = None
+    ) -> str:
+        """Convert an FS path to a Synapse ID
+
+        Args:
+            path (str): Path to a resource on the filesystem
+            starting_entity (str): Synapse ID for where to
+                start the traversal. Defaults to the root.
+
+        Returns:
+            str: Synapse ID for a file, folder, or project
+
+        Raises:
+            fs.errors.InvalidPath: If ``path`` is not absolute.
+            fs.errors.ResourceNotFound: If the ``path`` does
+                not resolve to existing entities.
+        """
+        original_path = path
+        starting_entity = starting_entity or self.root
+
+        if starting_entity is None:
+            starting_entity, _, path = path.partition(self.DELIMITER)
+            if not self.is_synapse_id(starting_entity):
+                message = (
+                    f"This SynapseFS is rootless, so the 1st part ({starting_entity}) "
+                    f"of every path ({original_path}) must be a Synapse ID."
+                )
+                raise ValueError(message)
+
+        current_entity = starting_entity
+        for part in path.strip(self.DELIMITER).split(self.DELIMITER):
+            if part in {"", self.DELIMITER, self.CWD_SYMBOL}:
+                continue
+
+            if part == self.PARENT_DIR:
+                current_entity = self._get_parent_entity(current_entity)
+                continue
+
+            children_list = self._get_children(current_entity)
```

Comment:
Just to confirm, when `part` is at its final element, current entity should be the synapse id of the container right before it.  For instance:

```
my/example/here
```

When `part` is `here`, `current_entity` should be the Synapse id of `example` right?

Reply:
On the `self._get_children(current_entity)` line, that's correct. The `current_entity` will get updated with the Synapse ID corresponding to `part` a little further below on the `children[next_part]["id"]` line. 

That said, I agree that the fact that `current_entity` and `part` don't track together is confusing. I've renamed `part` to `next_part` so they don't seem like they should be in sync.